### PR TITLE
fixes #8, fixes #12 added OrderedFirstN, First, FirstN

### LINKED ISSUE #8: Implement Set.First and Set.FirstN
```go
func (Set) First(less func(a, b interface{}) bool) interface{}
func (Set) FirstN(n int, less func(a, b interface{}) bool) Set
```

Return the first element or a new `Set` from the `OrderedRangeN` (#12). Naturally, the resulting `Set` will be unordered, but the caller can chain `OrderedElements` on the end if they want sorted output.

### LINKED ISSUE #12: Implement (Set).OrderedRangeN
```go
func (s Set) OrderedRangeN(less func(a, b interface{}) bool) Iterator
```

Iterates over the first _n_ elements per the given ordering. Doing this efficiently will require changing `s.root.orderedIterator` to `s.root.orderedIteratorN`, which bounds the size of the internal heap. To support the current unbounded behaviour, replace n = -1 with n = max-int.

### PR DESCRIPTION
**Pull request by nofun97**
Fixes #8 and #12 

Added:
1. First: takes the first element in a defined order
2. FirstN: returns a set of the first N elements in a defined order
3. OrderedRangeN: returns an array of n elements in a defined order

### CODE REVIEW

File: set_test.go [hunk 1]
```diff
@@ -208,6 +208,54 @@ func TestSetEqual(t *testing.T) {
...
+	assert.Equal(t, 0, s.First(less))
+
+	less = func(a, b interface{}) bool { return a.(int) > b.(int) }
+	assert.Equal(t, 1<<12-2, s.First(less))
+}
+
+func TestSetFirstN(t *testing.T) {
+	t.Parallel()
+
+	less := func(a, b interface{}) bool { return a.(int) < b.(int) }
```
Comment (ChloePlanet) [inline set_test.go]: less := Less(func(a, b interface{}) bool { return a.(int) < b.(int) })

File: set_test.go [hunk 2]
```diff
@@ -208,6 +208,54 @@ func TestSetEqual(t *testing.T) {
 	}
 }
 
+func TestSetFirst(t *testing.T) {
+	t.Parallel()
+
+	var s Set
+	less := func(a, b interface{}) bool { return a.(int) < b.(int) }
```
Comment (ChloePlanet) [inline set_test.go]: less := Less(func(a, b interface{}) bool { return a.(int) < b.(int) })

File: set_test.go [hunk 3]
```diff
@@ -208,6 +208,54 @@ func TestSetEqual(t *testing.T) {
...
+	assert.True(t, s.Equal(s.FirstN(1<<12-1, less)))
+
+	s = Iota(5)
+	assert.True(t, NewSet(0, 1, 2, 3, 4).Equal(s.FirstN(10, less)))
+}
+
+func TestSetOrderedFirstN(t *testing.T) {
+	t.Parallel()
+
+	less := func(a, b interface{}) bool { return a.(int) < b.(int) }
```
Comment (ChloePlanet) [inline set_test.go]: less := Less(func(a, b interface{}) bool { return a.(int) < b.(int) })

File: set_test.go [hunk 4]
```diff
@@ -208,6 +208,54 @@ func TestSetEqual(t *testing.T) {
 	}
 }
 
+func TestSetFirst(t *testing.T) {
+	t.Parallel()
+
+	var s Set
+	less := func(a, b interface{}) bool { return a.(int) < b.(int) }
+	assert.Panics(t, func() { s.First(less) })
```
Comment (ChloePlanet) [inline set_test.go]: Would be better: assert.Panics(t, func() { s.First(less) }, "empty set")

File: set_test.go [hunk 5]
```diff
@@ -208,6 +208,54 @@ func TestSetEqual(t *testing.T) {
...
+
+	s = Iota(1<<12 - 1)
+	assert.Equal(t, 0, s.First(less))
+
+	less = func(a, b interface{}) bool { return a.(int) > b.(int) }
```
Comment (ChloePlanet) [inline set_test.go]: Ditto, Less(...)

Comment (nofun97) [reply set_test.go]: whoops added all the changes

File: set_test.go (hunk 4 above)
Comment (nofun97) [reply set_test.go]: added the message

**Review by ChloePlanet — APPROVED** [review]

**Review by ChloePlanet — APPROVED** [review]

**Pull request merged**